---
title: "FluViz Tutorial"
author: "Molly Pratt"
date: "28/08/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction to FluViz
FluViz is a visualization tool for the [Influenza Classification Suite](https://doi-org.ezproxy.cscscience.ca/10.1111/irv.12722), which rapidly analyzes influenza virus sequences including clade mapping and antigenic site extraction. Example data for this tutorial including sequences from [GISAID](https://www.gisaid.org/) is available on [GitHub](https://github.com/mopratt/fluviz/tree/master/example_data/visualization)   
**Prior to visualizing your own data with FluViz, please use the Classification Suite workflow on Galaxy  to analyze and classify your influenza sequences.**  
  
**Influenza Classification Suite Workflow:**
![](FluViz/RMDs/flu-suite-wf.jpg)  
    
##### To visualize the output of the Classification Suite, only 3 R functions are needed:  
1. **combine_metadata()**  
2. **metamap()** *and*  
3. **snpplot()  **  
  
The source code for these R functions is available in the **rcode** folder from the GitHub page. If you want to follow this tutorial in RStudio, download the RScripts for each function, open and run them in RStudio, then download the example data and follow along:
    
### Combining metadata from GISAID and Line List output:  
The **combine_metadata()** function takes as input:  
1. the path to a metadata file (in .xls format) from GISAID  
2. the path to a line list file (in .csv format) from the Influenza Classification Suite  
3. the name of the column in the GISAID metadata that corresponds to the sequence labels.  
(*Note: this is likely going to be either "Isolate_Name" or "Isolate_Id"*)  
The output is a .csv file which is saved to your working directory. This .csv file can then be imported as a dataframe and further manipulated in R using common [**dplyr**](https://github.com/tidyverse/dplyr) verbs such as **mutate()** and **summarise()** in order to make the data more suitable for visualization. Another free tool for data cleaning is [Open Refine](https://openrefine.org/).    
```{r include=FALSE}
library(tidyverse)
library(readxl)
library(readr)
combine_metadata <- function(metadata, line_list, label_col, filename) {
  metadata_raw <- read_excel(metadata)
  metadata <- metadata_raw %>% select(label = all_of(label_col), Subtype, Location,
                                      Host, Host_Age, Host_Gender, Human_Specimen_Source,
                                      Animal_Specimen_Source, Collection_Date,
                                      Adamantanes_Resistance_geno,
                                      Oseltamivir_Resistance_geno, Zanamivir_Resistance_geno, 
                                      Peramivir_Resistance_geno, Other_Resistance_geno, 
                                      Adamantanes_Resistance_pheno, Oseltamivir_Resistance_pheno,
                                      Zanamivir_Resistance_pheno, Peramivir_Resistance_pheno,
                                      Other_Resistance_pheno, 
                                      Patient_Status, Outbreak)
  metadata <- metadata %>% mutate(label = str_replace_all(label, " ", "_")) %>% arrange(label)
  linelist_raw <- read_csv(line_list, skip = 3)
  linelist <- linelist_raw %>% select(label = `Sequence Name`, 
                                      `Clade`, 
                                      Num.aa.sub = `Number of Amino Acid Substitutions in Antigenic Sites`, 
                                      Percent.id = `% Identity of Antigenic Site Residues`)
  metadata_combined <- left_join(metadata, linelist, by = "label") %>% distinct()
  write_csv(metadata_combined, filename)
}
```
```{r, error=FALSE, warning=FALSE, message=FALSE}
combine_metadata(metadata = "example_data/visualization/gisaid_global_metadata.xls",
                 line_list = "example_data/visualization/line_list_global.csv",
                 label_col = "Isolate_Name",
                 filename = "combined_metadata.csv")

metadata <- read_csv("combined_metadata.csv")
# Optional steps: now is a good time to clean up the metadata and add any additional columns you would like to visualize. 
# Example:
metadata_clean <- metadata %>% 
  mutate(Year = substr(Collection_Date, 1, 4)) %>%
  separate(col = Location, c("Continent", "Country", "Region"), sep = "/")
```
  
### Reading in a phylogenetic tree file as a tree object using **treeio**  
Depending on the type of extension your tree file is saved as, remove the *"#"* from the appropriate line of code below and insert the path to your tree file:  
*If your tree extension is not listed below, see [Chapter 1: Importing Tree with Data](http://yulab-smu.top/treedata-book/chapter1.html) from Data Integration, Manipulation and Visualization of Phylogenetic Trees by Guangchuang Yu, PhD, the author of ggtree.*
```{r message=FALSE}
library(treeio)
library(ggtree)
#tree <- read.newick(path/to/tree.nwk)
#tree <- read.nexus(path/to/tree.nexus)
tree <- read.nhx("example_data/visualization/fasttree_global.nhx")
```
  
### Generate a cladogram of sequences and metadata heatmap:  
The **metamap()** function takes as input:  
1. a tree object, imported to R using **treeio** *(above)*    
2. a dataframe of metadata containing columns you would like to visualize in the heatmap  
3. a list of column names that will be visualized next to the cladogram. These columns can be categorical or numerical variables. 
The output is a cladogram of sequences (without branch length scaling) and a heatmap of the specified metadata. 
```{r include=FALSE}
library(ggplot2)
library(ggnewscale)
metamap <- function(tree, metadata, cols = "Clade") {
  met <- metadata %>% select(cols) %>% data.frame()
  treet <- tree %>% as.phylo() %>% as_tibble()
  treet_data <- treet %>% left_join(metadata, by = 'label')
  tree <- as.treedata(treet_data) %>% ggtree(branch.length = "none") + geom_tiplab(size = 2)
  rownames(met) <- metadata$label # metadata file must have first column as labels
  low = "red"
  high = "green"
  for (i in colnames(met)) {
    assign(paste0(i, ".df"), select(met, i))}
  for (i in colnames(met)) {
    if (i == first(colnames(met))) {
      h = 1
      ofs = 3
      let = 4
      x = 26.5
      if (is(get(paste0(i, ".df"))[,1], "numeric")) {
      assign(paste0("h", h), gheatmap(tree, get(paste0(i, ".df")),
                                      offset = ofs, width = 0.1, 
                                      colnames_position = "top", 
                                      colnames_offset_y = 1, 
                                      font.size = 3) +
               scale_fill_gradient(low = low, high = high, na.value = NA, name = i))
      } else {
        assign(paste0("h", h), gheatmap(tree, get(paste0(i, ".df")),
                                        offset = ofs, width = 0.1, 
                                        colnames_position = "top", 
                                        colnames_offset_y = 1, 
                                        font.size = 3) +
                 scale_fill_viridis_d(option = LETTERS[let], name = i))
      }
    } else {
      h = h + 1
      assign((paste0("h", h)), get(paste0("h", h - 1)) + new_scale_fill())
      h = h + 1
      ofs = ofs + 2
      let = if_else(let == 1, 4, let - 1)
      x = x + 6
      if (is(get(paste0(i, ".df"))[,1], "numeric")) {
      assign(paste0("h", h), gheatmap(get(paste0("h", h - 1)), get(paste0(i, ".df")), 
                                      offset = ofs, width = 0.1,
                                      colnames_position = "top", 
                                      colnames_offset_y = 1, 
                                      font.size = 3) +
             scale_fill_gradient(low = low, high = high, na.value = NA, name = i))
      }
      else {
        assign(paste0("h", h), gheatmap(get(paste0("h", h - 1)), get(paste0(i, ".df")), 
                                            offset = ofs, width = 0.1,
                                            colnames_position = "top", 
                                            colnames_offset_y = 1, 
                                            font.size = 3) +
                   scale_fill_viridis_d(option = LETTERS[let], name = i))
        }
    }
  }
  get(paste0("h", h)) + theme(legend.position = "bottom") + coord_cartesian(clip = "off")
}
```
```{r fig.height=10, fig.width=16, message=FALSE, warning=FALSE}
metamap(tree, metadata, cols = c("Clade", "Host", "Percent.id")) # specify columns from the metadata to be mapped
```
  
*Note: you can modify plot elements including the theme and legend placement by adding theme elements to replace the current theme, for example:*
``` {r fig.height=10, fig.width=16, message=FALSE, warning=FALSE}
metamap(tree, metadata, cols = c("Clade", "Host", "Percent.id")) + theme(legend.position = "left") # change the placement of the legend from the bottom of the plot to the left side of the plot
```

Now try mapping other columns, with and without data cleaning steps:
```{r message=FALSE, warning=FALSE, fig.height=10, fig.width=16}
metamap(tree, metadata, cols = c("Collection_Date", "Location")) + theme(legend.position = "right")
metamap(tree, metadata_clean, cols = c("Year", "Continent")) + theme(legend.position = "right")
```  
   
### Generate an interactive SNP plot based on the line list output: 
The **snpplot()** function takes as input:  
1. a tree object  
2. the path to a line list file (in .csv format) from the Influenza Classification Suite   
3. a dataframe of the sequence metadata  
4. the path to a clade definition file (in .csv format) that was used as input to the Influenza Classification Suite.  
  
The output is an interactive SNP plot of antigenic amino acids that differ from the reference sequence (provided to the Classification Suite). The amino acids are coloured according to whether they match the clade definition file. The tooltip information is specified within the **snpplot()** function, and can be easily modified.   
```{r include=FALSE}
library(plotly)
snpplot <- function(tree, line_list, metadata, clade_def) {
  line_list_raw <- read_csv(line_list)
  line_list <- line_list_raw %>%  distinct() %>% select(-2, -4) %>% 
    rename(label = X1, clade = X3) %>%
    slice(-2)
  tree_plot <- tree %>% as.phylo() %>% ggtree(branch.length = "none")
  tip_data <- tree_plot$data %>% arrange(desc(y)) %>% filter(isTip == TRUE)
  tip_labs <- tip_data %>% select(label, y)
  ref_align <- line_list %>% 
    dplyr::select(-'clade') %>%
    dplyr::semi_join(tip_data, by = "label")
  ref_align_fix <- t(ref_align)
  aa_rows <- ref_align_fix[1,]
  ref_align_fix <- ref_align_fix[-1,]
  colnames(ref_align_fix) <- aa_rows
  gapChar <- "."
  align_pos <- t(ref_align_fix)
  lalign_pos <- apply(align_pos, 1, function(x) {
  x != gapChar
  })
  lalign_pos <- as.data.frame(lalign_pos)
  lalign_pos$pos <- as.numeric(rownames(lalign_pos))
  lalign <- tidyr::gather(lalign_pos, label, value, -pos)
  aa_snp_data <- lalign[lalign$value, c("label", "pos")]
  ref <- line_list[1,] %>% 
    select(3:length(line_list)) %>%
    t() %>% 
    as.data.frame() %>%
    rownames_to_column(var = "pos") %>% 
    rename(ref.ID = V1)
  ref$pos <- as.numeric(as.character(ref$pos))
  aa_snps <- ref_align %>%
    gather(key = pos, value = snp.ID, -label) %>%
    mutate(pos = as.numeric(pos)) %>%
    right_join(aa_snp_data, by = c('label', 'pos')) %>%
    left_join(ref, by = "pos")
  clades <- read_csv(clade_def, col_names = FALSE)
  sites <- (length(clades) -2) / 2
  cnames <- rep(c("pos", "aa"), times = sites)
  cindex <- rep(c(1:sites), each = 2) 
  conames <- paste(cnames, cindex, sep = "-")
  conames <- c("clade", "depth", conames)
  colnames(clades) <- conames
  clades_long <- clades %>% pivot_longer(3:length(clades), names_to = c(".value", "set"),
                                         names_sep = "-",
                                         values_drop_na = TRUE) %>%
    dplyr::select(-3)
  clade_assign <- line_list %>% select('label', 'clade')
  aa_clades <- aa_snps %>% left_join(clade_assign, by = 'label') %>% 
    left_join(clades_long, by = c('clade', 'pos')) %>% 
    mutate(clade.def.pos = if_else(is.na(match(paste0(.$clade, .$pos),
                                               paste0(clades_long$clade, clades_long$pos))), FALSE, TRUE)) %>%
    rename(expected.aa = 'aa') %>%
    select(-'depth') %>% left_join(tip_labs, by = "label")
  aacols1 <-data.frame(aa = c("G", "A", "V", "L", "I", "P"), group = "nonpolar")
  aacols2 <-data.frame(aa = c("S", "T", "C", "M", "N", "Q"), group = "polar")
  aacols3 <-data.frame(aa = c("F", "W", "Y"), group = "aromatic")
  aacols4 <-data.frame(aa = c("K", "R", "H"), group = "basic")
  aacols5 <-data.frame(aa = c("D", "E"), group = "acidic")
  aacols6 <-data.frame(aa = c("X"), group = "unknown aa or clade")
  aacols <- aacols1 %>% bind_rows(aacols2, aacols3, aacols4, aacols5, aacols6)
  aa_clades_cols <- aa_clades %>% left_join(aacols, by = c("snp.ID" = "aa")) %>%
    rename(snp.group = group) %>%
    #mutate(snp.group = as.factor(snp.group)) %>%
    left_join(aacols, by = c("expected.aa" = "aa")) %>%
    rename(exp.group = group) %>%
    #mutate(exp.group = as.factor(exp.group)) %>%
    mutate(flag = if_else(snp.group == exp.group, "none", "aa.class.change")) %>%
    mutate(flag = if_else(clade == "No_Match", "no.clade", flag)) %>%
    mutate(flag = if_else(ref.ID == "X", "unknown.ref", flag)) %>%
    mutate(flag = if_else(is.na(flag), "none", flag)) %>%
    mutate(flag = factor(flag, levels = c("aa.class.change", "no.clade", "unknown.ref", "none")))
  myColours <- c("#E41A1C", "#FFFF33", "#984EA3", "#377EB8")
  names(myColours) <- levels(aa_clades_cols$flag)
  colScale <- scale_colour_manual(name = "flag", values = myColours)
  a <- aa_clades_cols %>% 
    ggplot(aes(x = pos, y = y)) + geom_point(aes(shape = 22, color= flag, 
                                                    text = paste("Label:", label, "\n",
                                                                 "Clade Assignment:", clade, "\n",
                                                                 "Position:", pos, "\n",
                                                                 "Identity:", snp.ID, "\n", 
                                                                 "Reference ID:", ref.ID, "\n",
                                                                 "Canonical Clade Site:", clade.def.pos, "\n",
                                                                 "Expected AA in Clade:", expected.aa)), 
                                                shape = 15, size = 1) + 
    scale_x_continuous(breaks = seq(0, 320, 20)) +
    theme_classic() + colScale + theme(axis.text.y=element_blank(),
                                       axis.ticks.y = element_blank(),
                                       axis.title.y = element_blank()) +
    labs(title = "Antigenic Amino Acid Substitutions", x = "position", color = "Amino Acid Group")
  snp_plot <- ggplotly(a, tooltip = "text")
  snp_plot
}
```
```{r fig.height=10, fig.width=16, message=FALSE, warning=FALSE}
snpplot(tree, line_list = "example_data/visualization/line_list_global.csv",
        metadata, clade_def = "example_data/visualization/FluA_H3_clade_defs.csv")
```
  
Now you should be able to generate plots using the fluviz functions **combine_metadata()**, **metamap()**, and **snpplot()**.

## Happy Visualizing!